addons/account/models/account.py

```diff
@@ -1375,29 +1375,40 @@ def _check_repartition_lines(self, lines):
+                    'repartition_type': 'tax',
+                    'factor_percent': 100.0,
+                    'company_id': base_line.company_id.id,
+                    'country_id': base_line.country_id.id,
+                }
+                self.env['account.tax.repartition.line'].create([
+                    {**values, 'invoice_tax_id': self.id},
+                    {**values, 'refund_tax_id': self.id},
+                ])
+            else:
+                raise ValidationError(_("Invoice and credit note repartition should have at least one tax repartition line."))
 
     @api.constrains('invoice_repartition_line_ids', 'refund_repartition_line_ids')
     def _validate_repartition_lines(self):
         for record in self:
-            invoice_repartition_line_ids = record.invoice_repartition_line_ids.sorted()
-            refund_repartition_line_ids = record.refund_repartition_line_ids.sorted()
-            record._check_repartition_lines(invoice_repartition_line_ids)
-            record._check_repartition_lines(refund_repartition_line_ids)
+            record._check_repartition_lines(record.invoice_repartition_line_ids)
+            record._check_repartition_lines(record.refund_repartition_line_ids)
 
-            if len(invoice_repartition_line_ids) != len(refund_repartition_line_ids):
+            # check consistency between invoice an refund lines
+            if len(record.invoice_repartition_line_ids) != len(record.refund_repartition_line_ids):
                 raise ValidationError(_("Invoice and credit note repartition should have the same number of lines."))
-
-            if not invoice_repartition_line_ids.filtered(lambda x: x.repartition_type == 'tax') or \
-                    not refund_repartition_line_ids.filtered(lambda x: x.repartition_type == 'tax'):
-                raise ValidationError(_("Invoice and credit note repartition should have at least one tax repartition line."))
-
-            index = 0
-            while index < len(invoice_repartition_line_ids):
-                inv_rep_ln = invoice_repartition_line_ids[index]
-                ref_rep_ln = refund_repartition_line_ids[index]
-                if inv_rep_ln.repartition_type != ref_rep_ln.repartition_type or inv_rep_ln.factor_percent != ref_rep_ln.factor_percent:
-                    raise ValidationError(_("Invoice and credit note repartitions should match (same percentages, in the same order)."))
-                index += 1
+            if any(
+                inv_rep_ln.repartition_type != ref_rep_ln.repartition_type
+                or inv_rep_ln.factor_percent != ref_rep_ln.factor_percent
+                for inv_rep_ln, ref_rep_ln in zip(
+                    record.invoice_repartition_line_ids.sorted(),
+                    record.refund_repartition_line_ids.sorted(),
+                )
+            ):
+                raise ValidationError(_("Invoice and credit note repartitions should match (same percentages, in the same order)."))
```

Comment:
By the way, if this is constrained... Why are we not just having "invoice tags" and "refund tags" on the repartition lines?
It's the only allowed difference, and it would simplify so many things.

Reply:
if I'm not mistaken, the only diff that matters for the issue is the one of _check_repartition_lines. Here it's pure refactoring, and our stable policy prevents any unnecessary refactoring because we might introduce bugs doing so.

So please, can you keep the diff minimal on stable versions?